documentation/man/features/kw-signature.rst

```diff
@@ -0,0 +1,94 @@
+===================================================
+kw-signature - Write commit and patch trailer lines
+===================================================
+
+.. _signature-doc:
+
+SYNOPSIS
+========
+| *kw* *signature* (-s(<string>) | \--add-signed-off-by=(<string>)) [\--verbose] [<sha> | <patch>]
+| *kw* *signature* (-r(<string>) | \--add-reviewed-by=(<string>)) [\--verbose] [<sha> | <patch>]
+| *kw* *signature* (-a(<string>) | \--add-acked-by=(<string>)) [\--verbose] [<sha> | <patch>]
+| *kw* *signature* (-f(<sha>) | \--add-fixes=(<sha>)) [\--verbose] [<sha> | <patch>]
+
+DESCRIPTION
+===========
+This is a wrapper to some useful usage of the 'git commit --amend',
+'git rebase' and the 'git interpret-trailers' commands. This kw command
+is able to perform usual operations over trailer lines of commits and patches.
+By default, it uses the commit currently pointed by ``HEAD`` as the target
+of the operation if user does not specify one. Also, no operation is used by
+default. At least one operation option must be given.
+
+This command does not allow repetition of options. For instance, the
+following command will give an error::
+
+  $ kw signature -r"Reviewer Name <mail>" -r"Some Name <mail>"
+
+However, using different options combined is possible. For instance,
+the following command will add **Reviewed-by** and **Signed-off-by**, in
+this order, to the commit's trailers pointed by ``HEAD`` using the
+user's git config::
+
+  $ kw signature -r -s
+
+Changing the order of the options affects the order the lines are written.
```

Comment:
Btw, google for this kernel page: Documentation/process/submitting-patches.rst. We can enforce the order suggested by the documentation.

Reply:
> I'm very confused about how exactly such order should be enforced. The documentation states that the **Signed-off-by** chain should represent chronologically the real route a patch took as it was propagated. Enforcing an order does not sound feasible to me because it would require the feature to predict the route and interactions between submitters and maintainers.
> 

We don't need to be perfect - maybe just close enough. With this idea in mind, we can try to create a heuristic for dealing with the sequence. For example, you can assume that the dev using the `kw signature` at the moment is the last person in the SoB chain.

> However, there're some details I've noticed while reading this document. It says that every `Co-developed-by` tag indicates authorship, which means that this tag must be immediately followed by a `Signed-off-by` of the same co-author. Since this a general rule, is it ok to implement it such that the option `--add-co-developed-by | -C` also adds an immediate `Signed-off-by` tag? Like:
> 
> ```shell
> $ kw signature -C'Developer Name <developer@mail.xyz>'
> # Written tags
> # Co-developed-by: Developer Name <developer@mail.xyz>
> # Signed-off-by: Developer Name <developer@mail.xyz>
> ```
> 
> Or would using `kw signature -C'blabla' -s'blabla'` still be preferable? 

The option added in your first example is the best way to handle the Co-developed-by tag. However, you might use the second example that you shared, and I think it should not be a problem as well. You just need to handle the duplicated Signed-off-by. 

>Also, it seems that the `Reported-by` flag should be followed by a `Closes` or `Link` tag if the bug report is available in the web. Should this somehow be passed via some argument in the command line?
> 

Keep in mind that in kw, we want to enforce the kernel way of doing things as much as possible. In this sense, we want to incentivize the users to follow the rules, which means we want them to use the Link/Close; however, from my experience, it is common not to have such a link. In this case, we need to define a heuristic. As a first suggestion, we can have something like this:

1. Reported-by might be used with a different (and optional) syntax. Something like this: `kw signature --reported-by "Name <email>;[link|close]=LINK"[,Name <email>...]`.
2. If the user does not provide a link: `kw signature --reported-by "Name <email>"` you can prompt a question asking if they have a link and give the option to insert it in case they forget. If they don't have a link, it should be ok. We probably want to have a force option to avoid asking.

> Sorry if I'm too lost with this. I've been thinking of how to enforce an order according to the documentation but I can't come up with a good idea right now.